src/main/java/org/apache/ibatis/builder/annotation/ProviderSqlSource.java

```diff
@@ -116,33 +118,45 @@ private SqlSource createSqlSource(Object parameterObject) {
     try {
       int bindParameterCount = providerMethodParameterTypes.length - (providerContext == null ? 0 : 1);
       String sql;
-      if (providerMethodParameterTypes.length == 0) {
+      if (parameterObject instanceof Map) {
+        if (bindParameterCount == 1 && providerMethodParameterTypes[0] == Map.class) {
+          sql = invokeProviderMethod(extractProviderMethodArguments(parameterObject));
+        } else {
+          @SuppressWarnings("unchecked")
+          Map<String, Object> params = (Map<String, Object>) parameterObject;
+          sql = invokeProviderMethod(extractProviderMethodArguments(params, providerMethodArgumentNames));
+        }
+      } else if (providerMethodParameterTypes.length == 0) {
         sql = invokeProviderMethod();
-      } else if (bindParameterCount == 0) {
-        sql = invokeProviderMethod(providerContext);
-      } else if (bindParameterCount == 1
-           && (parameterObject == null || providerMethodParameterTypes[providerContextIndex == null || providerContextIndex == 1 ? 0 : 1].isAssignableFrom(parameterObject.getClass()))) {
+      } else if (providerMethodParameterTypes.length == 1) {
+        if (providerContext == null) {
+          sql = invokeProviderMethod(parameterObject);
+        } else {
+          sql = invokeProviderMethod(providerContext);
+        }
+      } else if (providerMethodParameterTypes.length == 2) {
         sql = invokeProviderMethod(extractProviderMethodArguments(parameterObject));
-      } else if (parameterObject instanceof Map) {
-        @SuppressWarnings("unchecked")
-        Map<String, Object> params = (Map<String, Object>) parameterObject;
-        sql = invokeProviderMethod(extractProviderMethodArguments(params, providerMethodArgumentNames));
       } else {
-        throw new BuilderException("Error invoking SqlProvider method ("
-                + providerType.getName() + "." + providerMethod.getName()
-                + "). Cannot invoke a method that holds "
-                + (bindParameterCount == 1 ? "named argument(@Param)" : "multiple arguments")
-                + " using a specifying parameterObject. In this case, please specify a 'java.util.Map' object.");
+        throw new BuilderException("Cannot invoke SqlProvider method '" + providerMethod
+          + "' with specify parameter '" + (parameterObject == null ? null : parameterObject.getClass())
+          + "' because SqlProvider method arguments for '" + mapperMethod + "' is an invalid combination.");
```

Comment:
Shouldn't we use `Class#getName()` and `Method#getName()` ?
This might be a matter of preference, so I don't mind if you prefer `toString()`.

Reply:
In this case, I prefer to include a method arguments information(e.g. argument count and types) in exception message. Therefore I used `toString` as the easiest(simply) way to include that information. Should change to other way?

Reply:
Ah, I see. That makes sense.